### PR TITLE
ENG-2740 Use elapsed time for SSH retry timeout

### PR DESCRIPTION
**Pull request by fabgo**
Use elapsed time to determine the SSH retry timeout instead of the number of retries.

### CODE REVIEW

File: src/plugins/aws/ssh.ts [hunk 1]
```diff
@@ -22,11 +22,10 @@ import {
   AwsSshRoleRequest,
 } from "./types";
 
-/** Maximum number of attempts to start an SSH session
- *
- * Each attempt consumes ~ 1 s.
+/**
+ * It can take up to 1 minute for access to propagate on AWS, so set the time limit to 10 minutes.
```
Comment (nbrahms) [inline src/plugins/aws/ssh.ts:26]: What's the motivation to change from 30 seconds to 10 minutes?

I'd be surprised if any end user is willing to wait 10 minutes.

File: src/plugins/aws/ssh.ts [hunk 2]
```diff
@@ -22,11 +22,10 @@ import {
...
  */
-const MAX_SSH_RETRIES = 30;
+const TIME_LIMIT_MS = 10 * 60 * 1000;
```
Comment (nbrahms) [inline src/plugins/aws/ssh.ts:28]:
```suggestion
const TIME_LIMIT_MS = 60 * 1000;
```

Seems already long enough.

When I analyzed this previously, 99% + of provisioning finished within 8 seconds. Has that changed?

File: src/plugins/google/ssh.ts [hunk 3]
```diff
@@ -14,11 +14,10 @@ import { ensureGcpSshInstall } from "./install";
 import { importSshKey } from "./ssh-key";
 import { GcpSshPermissionSpec, GcpSshRequest } from "./types";
 
-/** Maximum number of attempts to start an SSH session
- *
- * The length of each attempt varies based on the type of error from a few seconds to < 1s
+/**
+ * It typically takes < 1 minute for access to propagate on GCP, so set the time limit to 2 minutes.
```
Comment (nbrahms) [inline src/plugins/google/ssh.ts:18]: Would be good to understand the CDF of successful attempts by time.

File: src/plugins/ssh/index.ts [hunk 4]
```diff
@@ -185,14 +185,11 @@ async function spawnSshNode(
...
 
-    const attemptsRemaining = options.attemptsRemaining;
     if (options.debug) {
       const gerund = options.isAccessPropagationPreTest
         ? "Pre-testing"
         : "Trying";
-      print2(
-        `Waiting for access to propagate. ${gerund} SSH session... (remaining attempts: ${attemptsRemaining})`
-      );
+      print2(`Waiting for access to propagate. ${gerund} SSH session...)`);
```
Comment (nbrahms) [inline src/plugins/ssh/index.ts:192]: Seems like we should display
```suggestion
      const remainingS = ((endTime - Date.now()) / 1e3 ).toFixed(1)
      print2(`Waiting for access to propagate. ${gerund} SSH session... (will wait up to ${remainingS} seconds)`);
```

**Review by nbrahms — COMMENTED** [review]
My biggest concern is why we changed from a 30 second to 10 minute (!) timeout for AWS access.

File: src/types/ssh.ts [hunk 5]
```diff
@@ -80,6 +80,8 @@ export type SshProvider<
 
   /** Unwraps this provider's types */
   requestToSsh: (request: CliPermissionSpec<PR, O>) => SR;
+  
+  timeLimit: number;
```
Comment (nbrahms) [inline src/types/ssh.ts:84]: This should have some jsdoc describing what it does. I might also suggest giving it a more descriptive name:

```suggestion
  /** Amount of time, in ms, to wait between granting access and giving up on attempting an SSH connection */
  propagationTimeoutMs: number;
```

Comment (fabgo) [reply src/types/ssh.ts:84]: Done

File: src/plugins/ssh/index.ts (hunk 4 above)
Comment (fabgo) [reply src/plugins/ssh/index.ts:134]: Done

File: src/plugins/aws/ssh.ts (hunk 1 above)
Comment (fabgo) [reply src/plugins/aws/ssh.ts:26]: The timeout is based on existing comments in ssh/index.ts:

AWS takes about 8 minutes, GCP takes under 1 minute to fully resolve access after it is granted.
During this time, calls to `aws ssm start-session` / `gcloud compute start-iap-tunnel`
will fail randomly with an various error messages.

See also this PR, which also mentions 10 minutes:
https://github.com/p0-security/p0cli/pull/19

Comment (fabgo) [reply src/plugins/aws/ssh.ts]: @gergas3 @nbrahms I can change the timeout if you want. Just let me know what to set it to.

Comment (gergas3) [reply src/plugins/aws/ssh.ts]: Was it a typo in https://github.com/p0-security/p0cli/pull/19 then? I didn't observe such a long propagation time for AWS. 30 seconds makes more sense.

GCP takes significantly longer, 1 minute seems about right but it's my guess would be it's not >99th percentile. We continue with 1m imo.

File: src/plugins/ssh/index.ts [hunk 6]
```diff
@@ -398,6 +394,6 @@ export const sshOrScp = async (args: {
     stdio: ["inherit", "inherit", "pipe"],
     debug: cmdArgs.debug,
     provider: request.type,
-    attemptsRemaining: sshProvider.maxRetries,
+    endTime: Date.now() + sshProvider.propagationTimeoutMs,
```
Comment (gergas3) [inline src/plugins/ssh/index.ts:397]: To globally wait max propagationTimeoutMs, we would have to the endTime once before calling `preTestAccessPropagationIfNeeded` (above), and pass the same endTime to `spawnSshNode` (here).

Now for a GCP sudo access the total propagationTimeoutMs is 2m * 2.

**Review by gergas3 — COMMENTED** [review]
With this change does this check in `plugins/ssh/index.ts`:
```
    if (
      match &&
      Date.now() <=
        beforeStart + (match.validationWindowMs || DEFAULT_VALIDATION_WINDOW_MS)
    ) {
      isEphemeralAccessDeniedException = true;
    }
```
still make sense? Do we want to have a timeout on individual subprocesses and consider the error non-ephemeral if we read it off the stderr too late.

I think it can be removed or set to a larger value because we have an overall cap on the time.

File: src/plugins/aws/ssh.ts (hunk 1 above)
Comment (fabgo) [reply src/plugins/aws/ssh.ts]: Changed it to 30 seconds.

File: src/plugins/ssh/index.ts (hunk 6 above)
Comment (fabgo) [reply src/plugins/ssh/index.ts]: Done

File: src/plugins/aws/ssh.ts (hunk 1 above)
Comment (nbrahms) [reply src/plugins/aws/ssh.ts]: The full info is that it takes AWS 30 minutes to get to 100% success rate. But we don't need to wait this long. We only need to wait until the chance that a single SSH attempt succeeds is high.

I'm not sure what we've changed in terms of delays, as those will affect this. But, when we were hammering it repeatedly in a loop, that took about 12 seconds.

**Review by gergas3 — APPROVED** [review]

**Pull request merged**